gateway-provider-security-shiro/src/main/java/org/apache/knox/gateway/shirorealm/KnoxCacheManager.java

```diff
@@ -17,11 +17,136 @@
  */
 package org.apache.knox.gateway.shirorealm;
 
-import org.apache.shiro.cache.ehcache.EhCacheManager;
+import org.apache.knox.gateway.ShiroMessages;
+import org.apache.knox.gateway.i18n.messages.MessagesFactory;
+import org.apache.shiro.ShiroException;
+import org.apache.shiro.cache.Cache;
+import org.apache.shiro.cache.CacheException;
+import org.apache.shiro.io.ResourceUtils;
+import org.apache.shiro.util.Destroyable;
+import org.apache.shiro.util.Initializable;
+import org.ehcache.CacheManager;
+import org.ehcache.config.CacheConfiguration;
+import org.ehcache.config.builders.CacheConfigurationBuilder;
+import org.ehcache.config.builders.CacheManagerBuilder;
+import org.ehcache.integrations.shiro.EhcacheShiro;
+import org.ehcache.xml.XmlConfiguration;
 
-public class KnoxCacheManager extends EhCacheManager {
+import java.net.MalformedURLException;
+import java.net.URL;
 
-  public KnoxCacheManager() {
-    setCacheManager(net.sf.ehcache.CacheManager.create());
+public class KnoxCacheManager implements org.apache.shiro.cache.CacheManager, Initializable, Destroyable {
+  private static final ShiroMessages LOG = MessagesFactory.get(ShiroMessages.class);
+
+  private org.ehcache.CacheManager manager;
+  private String cacheManagerConfigFile = "classpath:org/ehcache/integrations/shiro/ehcache.xml";
+  private boolean cacheManagerImplicitlyCreated;
+  private XmlConfiguration cacheConfiguration;
+
+  public CacheManager getCacheManager() {
+    return manager;
+  }
+
+  public void setCacheManager(CacheManager cacheManager) {
+    try {
+      destroy();
+    } catch (Exception e) {
+      LOG.errorClosingManagedCacheManager(e);
+    }
+    manager = cacheManager;
+    cacheManagerImplicitlyCreated = false;
+  }
+
+  public String getCacheManagerConfigFile() {
+    return cacheManagerConfigFile;
+  }
+
+  public void setCacheManagerConfigFile(String cacheManagerConfigFile) {
+    this.cacheManagerConfigFile = cacheManagerConfigFile;
+  }
+
+  @Override
+  public <K, V> Cache<K, V> getCache(String name) throws CacheException {
+    LOG.acquireEhcacheShiro(name);
+    try {
+      org.ehcache.Cache<Object, Object> cache = ensureCacheManager().getCache(name, Object.class, Object.class);
+
+      if (cache == null) {
+        LOG.noCacheFound(name);
+        cache = createCache(name);
+        LOG.ehcacheShiroAdded(name);
+      } else {
+        LOG.usingExistingEhcacheShiro(name);
+      }
+      return new EhcacheShiro<>(cache);
+    } catch (MalformedURLException | ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+      throw new CacheException(e);
+    }
+  }
+
+  private synchronized org.ehcache.Cache<Object, Object> createCache(String name)
```

Comment:
Why are we synchronizing?

Reply:
The implementation is based on the 2.x CacheManager. In our use-case there is no race condition currently however future changes might introduce. The synchronized block doesn't slow Knox's start-up because there is no contention.